### PR TITLE
Add validation to org subdomain during signup, and add a list of reserved subdomains.

### PR DESCRIPTION
**Pull request by dnephin**
## Summary

This PR adds validation to the org subdomain, including min/max length, character restrictions, and a list of reserved words that we should not allow to be subdomains.

Best viewed by individual commit, more details in commit messages.

### CODE REVIEW

File: api/signup.go [hunk 1]
```diff
@@ -12,6 +12,33 @@ type SignupOrg struct {
 	Subdomain string `json:"subDomain"`
 }
 
+var reservedSubDomains = []string{
```
Comment (mxyng) [inline api/signup.go:15]: @ssoroka also had a list of reserved subdomains. Perhaps you should work together on this

Comment (dnephin) [reply api/signup.go:15]: I'm happy to add or edit this list! I should have mentioned that in the PR description. 

Please do suggest changes to this list and I can update it.

File: api/passwordreset.go [hunk 2]
```diff
@@ -21,8 +21,13 @@ type VerifiedResetPasswordRequest struct {
...
 	return []validate.ValidationRule{
 		validate.Required("token", r.Token),
-		validate.String("token", r.Token, 10, 10, validate.AlphaNumeric...),
+		validate.StringRule{
+			Name:            "token",
+			Value:           r.Token,
+			MinLength:       10,
+			MaxLength:       10,
+			CharacterRanges: validate.AlphaNumeric,
+		},
```
Comment (ssoroka) [inline api/passwordreset.go:30]: way less elegant :P

File: api/signup.go [hunk 3]
```diff
@@ -12,6 +12,33 @@ type SignupOrg struct {
...
+	"grants", "connector", "login", "signup",
+	"system", "admin", "email", "bastion",
+}
+
+func (r SignupOrg) ValidationRules() []validate.ValidationRule {
+	return []validate.ValidationRule{
+		validate.Required("name", r.Name),
+		validate.Required("subDomain", r.Subdomain),
+		validate.ReservedStrings("subDomain", r.Subdomain, reservedSubDomains),
+		validate.StringRule{
```
Comment (ssoroka) [inline api/signup.go:27]: not necessarily a blocker, but it's missing the can't-start-with-dash case

**Review by ssoroka — APPROVED** [review]

File: api/signup.go (hunk 3 above)
Comment (dnephin) [reply api/signup.go:27]: I added first character validation to the `StringRule` and added it here as well

**Pull request merged**